src/PoseNet/index.js

```diff
@@ -22,80 +24,74 @@ const DEFAULTS = {
     this.detectionType = detectionType || DEFAULTS.detectionType;
     this.imageScaleFactor = options.imageScaleFactor || DEFAULTS.imageScaleFactor;
     this.outputStride = options.outputStride || DEFAULTS.outputStride;
     this.flipHorizontal = options.flipHorizontal || DEFAULTS.flipHorizontal;
     this.minConfidence = options.minConfidence || DEFAULTS.minConfidence;
     this.multiplier = options.multiplier || DEFAULTS.multiplier;
+    this.ready = callCallback(this.load(), callback);
+  }
 
-    posenet.load(this.multiplier)
-      .then((net) => {
-        this.net = net;
-        if (this.video) {
-          (this.video.onplay = () => {
-            if (this.detectionType === 'single') {
-              this.singlePose(callback);
-            } else if (this.detectionType === 'multiple') {
-              this.multiPose(callback);
-            }
-          })();
-        }
-      })
-      .catch((err) => { console.error(`Error loading the model: ${err}`); });
+  async load() {
+    const net = await posenet.load(this.multiplier);
+    this.net = net;
+    if (this.video) {
+      await new Promise((resolve) => {
+        this.video.onplay = resolve;
+      });
+      if (this.detectionType === 'single') {
+        return this.singlePose();
+      }
+      return this.multiPose();
+    }
+    return this;
   }
 
   skeleton(keypoints, confidence = this.minConfidence) {
     return posenet.getAdjacentKeyPoints(keypoints, confidence);
   }
 
   /* eslint max-len: ["error", { "code": 180 }] */
-  singlePose(inputOrCallback, cb = () => {}) {
+  async singlePose(inputOr) {
     let input;
-    let callback = cb;
 
-    if (inputOrCallback instanceof HTMLImageElement || inputOrCallback instanceof HTMLVideoElement) {
-      input = inputOrCallback;
-    } else if (typeof inputOrCallback === 'object' && (inputOrCallback.elt instanceof HTMLImageElement || inputOrCallback.elt instanceof HTMLVideoElement)) {
-      input = inputOrCallback.elt; // Handle p5.js image and video
-    } else if (typeof inputOrCallback === 'function' && this.video) {
+    if (inputOr instanceof HTMLImageElement || inputOr instanceof HTMLVideoElement) {
+      input = inputOr;
+    } else if (typeof inputOr === 'object' && (inputOr.elt instanceof HTMLImageElement || inputOr.elt instanceof HTMLVideoElement)) {
+      input = inputOr.elt; // Handle p5.js image and video
+    } else if (typeof inputOr === 'function' && this.video) {
       input = this.video;
-      callback = inputOrCallback;
     }
 
-    this.net.estimateSinglePose(input, this.imageScaleFactor, this.flipHorizontal, this.outputStride)
-      .then((pose) => {
-        callback([{ pose, skeleton: this.skeleton(pose.keypoints) }]);
-        tf.nextFrame().then(() => { this.singlePose(callback); });
-      });
+    const pose = await this.net.estimateSinglePose(input, this.imageScaleFactor, this.flipHorizontal, this.outputStride);
+    this.emit('pose', [{ pose, skeleton: this.skeleton(pose.keypoints) }]);
```

Comment:
What's the benefit of using `EventEmitter`? 

Although, looking at this class again, I think we should maintain consistency across all methods an have the user manage the loop events. So a working example of this should be:

```javascript
// Create the class with no `detectionType` argument
const poseNet = ml5.poseNet(video, modelLoaded);

// Make a prediction using the video
poseNet.singlePose((err, results) => {
  poses = results;
});
```

That way this class also matches the style of the other ml5 methods
```javascript
// Make a prediction using the an image
poseNet.singlePose(img, (err, results) => {
  poses = results;
});
```

Reply:
> What's the benefit of using EventEmitter?

An EventEmitter seems like the right abstraction for PoseNet because:

* It allows the user to both _listen_ (addEventListener/on) and _unlisten_ (removeEventListener/off) from new poses
* It allows multiple listeners to the poses

For example: let's say that you want to show both the poses overlaid on a video element, and those poses as JSON. With an event listener, you'd just call `.on(` with two different listeners. And if want to stop the video, you can use `.off(` to stop listening for poses.

That's why I don't think that this _should_ match the style of other ml5 methods, because it's doing something different. It's totally appropriate to use a callback or a promise if you're calling a method that's expected to have one result. But it's very unusual to use the callback / promise pattern for things that happen based on user input, like frames from a webcam. That's why, for example, we use events for click events on the web.

I think this API is pretty awkward even with the EventEmitter addition: for example, it's not possible to create a PoseNet instance, have it generate single poses, and then switch it to generating multi poses. It's stuck between this idea of immutable models - in which case, the answer would be 'just throw it away and create a new PoseNet', and models that support change - in which case we'd have a single `tf.nextFrame` loop that does an arbitrary 'current setting' on each frame. Currently it _looks_ like you'd be able to call singlePose() and multiPose() and change the PoseNet behavior, but in fact every time you call them you _add_ a new loop and extra attendant computation.